### DISCUSSION TITLE
Using Bing/Azure maps in iTwin viewer

### DISCUSSION THREAD

**Discussion by abdul-lashkarwala** [iTwin/community#64] [Q&A]
Hello team,

I am working on creating an iTwin web viewer using iTwin.js and would like to use Bing/Azure base maps in the app. As indicated in the documentation, I have created an Azure Maps account and added the following configuration in the Viewer and the respective key in the .env file, but it isn't working. Am I missing anything here? I would appreciate any help from the community.

mapLayerOptions={{
    AzureMaps: {
        key: "subscription-key",
        value: process.env.IMJS_AZURE_MAP_KEY ?? "",
    },

}}

**Comment by hl662** [comment]
Hello! Can I ask what was the source of the documentation? We haven't added support for AzureMaps yet, and Bing Maps is deprecated. There is a draft [PR](https://github.com/iTwin/itwinjs-core/pull/7010) up for Azure Maps support, but we haven't prioritized picking it up just yet, because Azure Maps is not feature parity with Bing Maps in terms of map imagery. We are instead however, looking to switch to using google maps: https://github.com/iTwin/itwinjs-core/pull/7414

**Reply by abdul-lashkarwala** [reply]
Hi @hl662 , thanks for your reply. I was referring to this documentation - https://www.itwinjs.org/reference/core-frontend/maplayers/maplayeroptions/

**Reply by hl662** [reply]
Thanks for sharing, we'll look into updating the documentation to recommend users to not use those...

**Comment by hl662** [comment]
At the moment, you can implement your own MapLayerProvider for AzureMaps - here is a [link](https://www.itwinjs.org/reference/core-frontend/maplayers/maplayerimageryprovider/) to our abstract interface for a MapLayerImageryProvider. Upon startup of the `IModelApp`, you can register your own provider, like in this [class](https://github.com/iTwin/itwinjs-core/blob/1949fe6a4244a2567446f042bf9c2cbc2310b4c1/extensions/map-layers-formats/src/mapLayersFormats.ts#L38).

Then, assuming you have a viewport created and accessible through your own iTwin viewer, you can do the following to change the map layer imagery for that viewport at any point (perhaps as early in the viewer initialization stage as you can) like so:

```ts
viewport.displayStyle.attachMapLayer({
      settings: ImageMapLayerSettings.fromJSON({name: "PhillyRailLines", formatId: "ArcGIS", url: "https://services7.arcgis.com/nZ2Vb4CUwdo9AIiQ/ArcGIS/rest/services/PhillyRailLines/MapServer"}), 
      mapLayerIndex: {isOverlay: true, index: 0}
    });
```

The formatId will be the custom formatId you decide for your own imagery provider. You can see how one of our colleagues is implementing a GoogleMaps Provider through this [link](https://github.com/iTwin/itwinjs-core/pull/7414/files#diff-70df4e940b6f79371a3c72101db8026a6f8a42cd71d33ef85ed9758850b95fc1R89), and where they set the map layer [here](https://github.com/iTwin/itwinjs-core/pull/7414/files#diff-00630f98539112d0c26a2fdc786d40831dbd8af303449cf0bcb3263e61deb755R478) (it might be a different approach to setting the map layer, but they still make use of the concept of attaching map layers)

**Reply by abdul-lashkarwala** [reply]
Hi @hl662 , thanks for your reply. I am new to iTwin.JS, can you share a working sample that I can refer to integrate basemaps?

**Reply by hl662** [reply]
Sure! In this [sample](https://www.itwinjs.org/sandboxes/iTwinPlatform/Car%20Particle%20Effect), is a working example of how to set a viewport display style's backgroundMapBase, to update the terrain being shown. If you navigate to the `CarDecorationApi.ts` file, at line 125, you'll see the snippet of code where that is performed. This sample is outdated, so that's why you'll see it still being able to use BingMaps - it's using a previous BingMap key created before deprecation.

Since you're new to iTwinJS, the concepts of Widgets and Providers may feel new and quite overwhelming. I recommend going through some of the top learning tutorials [here](https://developer.bentley.com/tutorials/), or you can look through the [samples](https://developer.bentley.com/samples/) others have created.
Will you be using React to implement your iTwin viewer?

**Reply by abdul-lashkarwala** [reply]
Thanks for your reply. Yes, I am going to use React for our iTwin viewer.
